src/libcore/ptr.rs

```diff
@@ -946,6 +946,73 @@ pub unsafe fn write_volatile<T>(dst: *mut T, src: T) {
     intrinsics::volatile_store(dst, src);
 }
 
+/// Freezes `count * size_of::<T>()` bytes of memory, converting uninitialized data into
+/// arbitrary but fixed data.
+///
+/// Uninitialized memory has undefined contents, and interaction with those contents
+/// can easily cause undefined behavior. Freezing the memory avoids those issues by
+/// converting the memory to an initialized state without actually needing to write to
+/// all of it. This function has no effect on memory which is already initialized.
+///
+/// While this function does not actually physically write to memory, it acts as if it
+/// does. For example, calling this method on data which is concurrently accessible
+/// elsewhere is undefined behavior just as it would be to use `ptr::write`.
+///
+/// # Warning
+///
+/// Take care when using this function as the uninitialized memory being frozen can
+/// contain bits and pieces of previously discarded data, including sensitive
+/// information such as cryptographic keys.
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following conditions are violated:
+///
+/// * `dst` must be [valid] for writes.
+///
+/// * Every bit representation of `T` must be a valid value.
```

Comment:
Why is that a precondition to `freeze`? `freeze` does not assert validity of anything.

`freeze` on a `*mut bool` is not UB. Just dereferencing the pointer later is.

Reply:
I thoughts that what you meant in this comment: https://github.com/rust-lang/rust/pull/58363#discussion_r255526327. If the validity requirements are less strict I can move this to be a more general warning about use of the function rather than a hard UB constraint.

Reply:
The phrasing here seems congruent with the behavior of `freeze::<*mut bool>(...)` tho. The value is the pointer (`*mut bool`), not the pointee (`bool`).

Reply:
The validity requirements come in once you actually operate on or a make a copy of a value at some type.  So the following is UB:
```rust
let x = MaybeUninit::<bool>::uninitialized();
ptr::freeze(x.as_mut_ptr());
let x = x.into_initialized(); // UB
```
But it's not the freeze that is UB, it's the `into_initialized`!

Reply:
@RalfJung Oh... you're talking about `*mut bool` as the type of the first argument, not `T == *mut bool`... i.e. `ptr::freeze<bool>(x.as_mut_ptr(): *mut bool)` rather than `ptr::freeze<*mut bool>(??: *mut *mut bool)`. As you have phrased it with `x.into_initialized()` being UB, this seems more of a post-condition?